### PR TITLE
Upgrade toolchain to nightly-2023-03-09

### LINKED ISSUE #2283: Update rust toolchain version for Kani 0.24
Biweekly update for the rust toolchain version. For Kani 0.24, we will target the new year version: nightly-2023-03-09.

### PR DESCRIPTION
**Pull request by qinheping**
### Description of changes: 

Upgrade our toolchain to `nightly-2023-03-09`. The changes here are related to the following changes:

- https://github.com/rust-lang/rust/pull/107163
- https://github.com/rust-lang/rust/pull/107838
- https://github.com/rust-lang/rust/pull/108364
- https://github.com/rust-lang/rust/pull/108505
- https://github.com/rust-lang/rust/pull/108029
- https://github.com/rust-lang/rust/pull/108047
- https://github.com/rust-lang/rust/pull/107753
- https://github.com/rust-lang/rust/pull/108250
- https://github.com/rust-lang/rust/pull/108020
- https://github.com/rust-lang/rust/pull/108112

### Resolved issues:

Resolves #2283

### Related RFC:

Optional #ISSUE-NUMBER.

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: kani-compiler/src/kani_middle/provide.rs [hunk 1]
```diff
@@ -6,10 +6,10 @@
 
 use crate::kani_middle::reachability::{collect_reachable_items, filter_crate_items};
 use crate::kani_middle::stubbing;
+use crate::kani_middle::ty::query::query_provided::collect_and_partition_mono_items;
```
Comment (zhassan-aws) [inline kani-compiler/src/kani_middle/provide.rs:9]: This change looks a bit odd. Isn't this importing `collect_and_partition_mono_items` from `rustc_middle`? If so, why is it referring to `crate::kani_middle`?

**Review by zhassan-aws — COMMENTED** [review]
Thanks for looking into this @qinheping!

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 2]
```diff
@@ -783,19 +783,31 @@ impl<'tcx> GotocCtx<'tcx> {
...
-        let param_env_and_layout = ty::ParamEnv::reveal_all().and(layout);
+        let param_env_and_layout = ty::ParamEnv::reveal_all().and(ty);
 
         // Then we check if the type allows "raw" initialization for the cases
         // where memory is zero-initialized or entirely uninitialized
-        if intrinsic == "assert_zero_valid" && !self.tcx.permits_zero_init(param_env_and_layout) {
+        if intrinsic == "assert_zero_valid"
+            && !self
+                .tcx
+                .check_validity_requirement((ValidityRequirement::Zero, param_env_and_layout))
```
[the remark this answers is not in the document]
Comment (qinheping) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:793]: The PR https://github.com/rust-lang/rust/pull/108505 later deprecated `check_validity_of_init`.

[the remark this answers is not in the document]
Comment (zhassan-aws) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs:793]: I see.

**Comment by tautschnig** [comment]
I have rebased the branch to include the CBMC 5.79.0 update to make sure this dependency is sorted out (but obviously we are currently failing well before reaching CBMC). I will look into those failures.

File: kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs [hunk 3]
```diff
@@ -783,19 +783,31 @@ impl<'tcx> GotocCtx<'tcx> {
             );
         }
 
-        let param_env_and_layout = ty::ParamEnv::reveal_all().and(layout);
+        let param_env_and_layout = ty::ParamEnv::reveal_all().and(ty);
```
[the remark this answers is not in the document]
Comment (tautschnig) [reply kani-compiler/src/codegen_cprover_gotoc/codegen/intrinsic.rs]: Done in an earlier PR.

**Pull request closed without merging**